File: solution/1463. Cherry Pickup II/1463_01.cpp
class Solution {
public:
    int cherryPickup(vector<vector<int>>& grid) {
        int h = grid.size(), w = grid[0].size();
        
        vector<vector<vector<int>>> cache(h, vector<vector<int>>(w, vector<int>(w, -1)));
        
        function<int(int, int, int)> dp = [&](int y, int x1, int x2) {
            if (y < 0 || y >= h || x1 < 0 || x1 >= w || x2 < 0 || x2 >= w) return 0;
            int& ans = cache[y][x1][x2];
            if (ans != -1) return ans;
            
            const int cur = grid[y][x1] + (x1 != x2) * grid[y][x2];
            for (int d1 = -1; d1 <= 1; ++d1)
                for (int d2 = -1; d2 <= 1; ++d2)
                    ans = max(ans, cur + dp(y + 1, x1 + d1, x2 + d2));
            
            return ans;
        };
        
        return dp(0, 0, w - 1);
    }
};

File: solution/1463. Cherry Pickup II/1463_01.py
class Solution:
    def cherryPickup(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        
        @lru_cache(None)
        def dfs(y, x1, x2):
            if y == m:
                return 0
            
            cherry = 0
            for i in range(max(0, x1-1), min(n-1, x1+1) + 1):
                for j in range(max(0, x2-1), min(n-1, x2+1) + 1):
                    cherry = max(cherry, dfs(y + 1, i, j))
            return cherry + grid[y][x1] + (x1 != x2) * grid[y][x2]
        
        return dfs(0, 0, n - 1)
    
